python/ray/tune/suggest/optuna.py

```diff
@@ -0,0 +1,149 @@
+import logging
+import pickle
+
+from ray.tune.result import TRAINING_ITERATION
+
+try:
+    hyperopt_logger = logging.getLogger("hyperopt")
+    hyperopt_logger.setLevel(logging.WARNING)
+    import optuna as ot
+except ImportError:
+    ot = None
+
+from ray.tune.suggest import Searcher
+
+logger = logging.getLogger(__name__)
+
+
+class _Param:
+    def __getattr__(self, item):
+        def _inner(*args, **kwargs):
+            return (item, args, kwargs)
+
+        return _inner
+
+
+param = _Param()
+
+
+class OptunaSearch(Searcher):
+    """A wrapper around Optuna to provide trial suggestions.
+
+    `Optuna <https://optuna.org/>`_ is a hyperparameter optimization library.
+    In contrast to other libraries, it employs define-by-run style
+    hyperparameter definitions.
+
+    This Searcher is a thin wrapper around Optuna's search algorithms.
+    You can pass any Optuna sampler, which will be used to generate
+    hyperparameter suggestions.
+
+    Please note that this wrapper does not support define-by-run, so the
+    serach space will be configured before running the optimization. You will
+    also need to use a Tune trainable (e.g. using the function API) with
+    this wrapper.
+
+    For defining the search space, use ``ray.tune.suggest.optuna.param``
+    (see example).
+
+    Args:
+        space (list): Hyperparameter search space definition for Optuna's
+            sampler. This is a list, and samples for the parameters will
+            be obtained in order.
+        metric (str): Metric that is reported back to Optuna on trial
+            completion.
+        mode (str): One of {min, max}. Determines whether objective is
+            minimizing or maximizing the metric attribute.
+        sampler (optuna.samplers.BaseSampler): Optuna sampler used to
+            draw hyperparameter configurations. Defaults to ``TPESampler``.
+
+    Example:
+
+    .. code-block:: python
+
+        from ray.tune.suggest.optuna import OptunaSearch, param
+
+        space = [
+            param.suggest_uniform("a", 6, 8),
+            param.suggest_uniform("b", 10, 20)
+        ]
+
+        algo = OptunaSearch(
+            space,
+            metric="loss",
+            mode="min")
+
```

Comment:
maybe also do a .. versionadded: 

https://www.sphinx-doc.org/en/master/usage/restructuredtext/directives.html#directive-versionadded

Reply:
That will be 0.8.8 then, right?

Reply:
yep! sounds good